<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.450">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Paweł Czyż">
<meta name="dcterms.date" content="2024-07-15">
<meta name="description" content="Essentially one. But some care is needed to count them properly.">

<title>Paweł Czyż - How many distinct Ising models are over there?</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
</style>


<script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<script src="../site_libs/quarto-html/quarto.js"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting-dark.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="dark">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>

  <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<link rel="stylesheet" href="../styles.css">
</head>

<body class="nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg navbar-dark ">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container">
    <a class="navbar-brand" href="../index.html">
    <span class="navbar-title">Paweł Czyż</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="../index.html" rel="" target="">
 <span class="menu-text">About</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../blog.html" rel="" target="">
 <span class="menu-text">Blog</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../publications.html" rel="" target="">
 <span class="menu-text">Publications</span></a>
  </li>  
</ul>
            <div class="quarto-navbar-tools ms-auto">
</div>
          </div> <!-- /navcollapse -->
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#a-different-ising-model" id="toc-a-different-ising-model" class="nav-link active" data-scroll-target="#a-different-ising-model">A “different” Ising model</a></li>
  <li><a href="#arent-these-the-same-model" id="toc-arent-these-the-same-model" class="nav-link" data-scroll-target="#arent-these-the-same-model">Aren’t these the same model?</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">How many distinct Ising models are over there?</h1>
</div>

<div>
  <div class="description">
    Essentially one. But some care is needed to count them properly.
  </div>
</div>


<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Paweł Czyż </p>
          </div>
  </div>
    
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">July 15, 2024</p>
    </div>
  </div>
  
    
  </div>
  

</header>

<p>Recall “the” <a href="https://en.wikipedia.org/wiki/Ising_model">Ising model</a>. We have a graph with <span class="math inline">\(G\)</span> nodes and each node takes a value <span class="math inline">\(s_g \in \{-1, 1\}\)</span>. It is parameterised by a symmetric <span class="math inline">\(G\times G\)</span> matrix <span class="math inline">\(\Omega = (\Omega_{gg'})\)</span>, representing interactions between different nodes, and an additional vector <span class="math inline">\(\mathbf{\alpha} = (\alpha_g)\)</span>. In physics, <span class="math inline">\(s_g\)</span> may be the magnetic moments of different sites in a lattice of magnets: <span class="math inline">\(\Omega\)</span> describes interactions between different sites (e.g., due to the spatial structure of the lattice it can be constrained) and <span class="math inline">\(\mathbf{\alpha}\)</span> describes the interactions with an external magnetic field, providing asymmetry between the states <span class="math inline">\(-1\)</span> and <span class="math inline">\(+1\)</span>.</p>
<p>Given the configuration vector <span class="math inline">\(\mathbf{s} = (s_g)\)</span>, the energy of the system is given as <span class="math display">\[
    E(\mathbf{s}) = \mathbf{s}^T \Omega \mathbf{s} + \mathbf{\mathbf{\alpha}}^T \mathbf{s} = \sum_{g} \alpha_g s_g + \sum_{g}\sum_{g'}  \Omega_{gg'}s_g s_{g'}.
\]</span></p>
<p>Given the energy, we will be interested in the probability distribution <span class="math display">\[
  P(\mathbf{s}) = \mathcal Z^{-1} \exp(-E(\mathbf{s})),
\]</span> where <span class="math inline">\(\mathcal Z\)</span> is the normalising constant and can depend on the parameters <span class="math inline">\(\mathbf{\alpha}\)</span> and <span class="math inline">\(\Omega\)</span>.</p>
<p>This is an interesting system: it can be specified by <span class="math inline">\(G(G+1)/2 + G = O(G^2)\)</span> parameters, but it <em>can</em> attribute distinct energies to each of the possible <span class="math inline">\(2^G\)</span> states. It does not have to be the case for every set of parameters, though: for example, if we restrict the parameters, e.g., set <span class="math inline">\(\mathbf{\alpha} = \mathbf{0}\)</span>, then the states <span class="math inline">\(\mathbf{s}\)</span> and <span class="math inline">\(\mathbf{-s}\)</span> have the same energy and, at most, <span class="math inline">\(2^{G-1}\)</span> distinct energy levels can be represented. Moreover, even if this model attributes <span class="math inline">\(2^G\)</span> distinct energies, it does not need to be a good model for an arbitrary probability distribution over the set <span class="math inline">\(\{-1, 1\}^G\)</span>, which can require even <span class="math inline">\(2^G\)</span> parameters to specify.</p>
<p>If we decided to count different energy functions <span class="math inline">\(E\)</span> as different models, the answer “how many of different Ising models do we have” would be <span class="math inline">\(+\infty\)</span>. However, this is not a very compelling answer and as all of them have the same structure, we will think of them as of one model.</p>
<p>However, not all the parameters are really necessary. Because for every <span class="math inline">\(s_g \in \{0, 1\}\)</span> we have <span class="math inline">\(s_g^2 = 1\)</span>, for every state <span class="math inline">\(\mathbf{s}\)</span> we have a summand <span class="math inline">\(\mathbf{s}^T \mathrm{diag}(\Omega) \mathbf{s} = \sum_g \Omega_{gg}\)</span>. This offset does not change energy differences and makes the statistical model unidentifiable. We can therefore drop the diagonal terms (by setting <span class="math inline">\(\Omega_{gg} = 0\)</span>) and think of the Ising model as of one having <span class="math inline">\(G(G-1) / 2 + G = G(G+1)/2\)</span> parameters. Note that this is <em>not</em> the same as dropping the <span class="math inline">\(\mathbf{\alpha}\)</span> terms, which are needed to introduce asymmetry between <span class="math inline">\(\mathbf{s}\)</span> and <span class="math inline">\(-\mathbf{s}\)</span> states.</p>
<section id="a-different-ising-model" class="level2">
<h2 class="anchored" data-anchor-id="a-different-ising-model">A “different” Ising model</h2>
<p>Now consider a physical model of interacting magnets, but a statistical model representing mutations. This idea dates back to the <a href="https://arxiv.org/abs/1208.3555">2013 paper of Lingzhou Xue, Hui Zou and Tianxi Cai</a> and the <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC1226070/">2001 paper of Jacek Majewski, Hao Li and Jurg Ott</a>, in which they use Ising models described above to model mutations.</p>
<p>However, for me assigning numbers from the set <span class="math inline">\(\{-1, 1\}\)</span> to mutation data is not natural: I think of a binary genotype in terms of a vector <span class="math inline">\(\mathbf{y}\in \{0, 1\}^G\)</span> representing the presence or the absence of a mutation at a particular locus.</p>
<p>The energy for that model would be <span class="math display">\[
  \tilde E(\mathbf{y}) = \sum_g \pi_g y_g + \sum_{g'\neq g} \Theta_{gg'} y_gy_g',
\]</span></p>
<p>where <span class="math inline">\(\pi_g\)</span> is then related to the probability of gene mutation in the model where all mutations arise independently (i.e., <span class="math inline">\(\Theta_{gg'} = 0\)</span> for all <span class="math inline">\(g\neq g'\)</span>. See the model we discussed <a href="../posts/discrete-intractable-likelihood.html">in this blog post</a>) and <span class="math inline">\(\Theta_{gg'}=\Theta_{g'g}\)</span> are used to model some (anti-)correlations between genes <span class="math inline">\(g\)</span> and <span class="math inline">\(g'\)</span>. Note that as <span class="math inline">\(y_g = y_g^2\)</span> for <span class="math inline">\(y_g \in \{0, 1\}\)</span> we can write <span class="math inline">\(\Theta_{gg} = \pi_g\)</span> and use a single matrix <span class="math inline">\(\Theta\)</span>: <span class="math display">\[
  \tilde E(\mathbf{y}) = \mathbf{y}^T \Theta \mathbf{y} = \sum_{g}\sum_{g'} \Theta_{gg'} y_g y_{g'}.
\]</span></p>
<p>This model can also result in <span class="math inline">\(2^G\)</span> different energy levels.</p>
</section>
<section id="arent-these-the-same-model" class="level2">
<h2 class="anchored" data-anchor-id="arent-these-the-same-model">Aren’t these the same model?</h2>
<p>We can hope that both models are essentially equivalent, i.e., if we relabel <span class="math inline">\(\mathbf{s}\)</span> to <span class="math inline">\(\mathbf{y}\)</span>, we can find the parameters <span class="math inline">\(\Theta\)</span> such that energy in the new model, <span class="math inline">\(\tilde E(\mathbf{y})\)</span> will “agree” with the old energy <span class="math inline">\(E(\mathbf{s})\)</span>. (And, that for every state <span class="math inline">\(\mathbf{y}\)</span> relabelled to <span class="math inline">\(\mathbf{s}\)</span>, we can find parameters <span class="math inline">\(\mathbf{\alpha}\)</span> and <span class="math inline">\(\Omega\)</span> such that the energies agree).</p>
<p>We shouldn’t hope for the <em>exact</em> equality of energies (after all recall that (a) zero point energy is not well defined in the first model, as we can always change the diagonal terms of <span class="math inline">\(\Omega\)</span> shifting all energy levels simultaneously by the same number, (b) only the energy differences matter, as in this model we can observe only frequencies of different states and shifting the energy levels doesn’t change the probability distribution), but the equality of the differences <span class="math inline">\(\tilde E(\mathbf{y}) - \tilde E(\mathbf{y}')\)</span> and <span class="math inline">\(E(\mathbf s) - E(\mathbf s')\)</span> whenever we respectively relabel <span class="math inline">\(\mathbf s\)</span> and <span class="math inline">\(\mathbf s'\)</span> to <span class="math inline">\(\mathbf y\)</span> and <span class="math inline">\(\mathbf y'\)</span>. (And that relabelling from <span class="math inline">\(\mathbf y\)</span> to <span class="math inline">\(\mathbf s\)</span> also “works”, in the sense explained above).</p>
<p>We can consider two relabelling procedures:</p>
<p><span class="math display">\[
  s_g = -1 \leftrightarrow y_g= 0,\quad s_g =1 \leftrightarrow y_g = 1
\]</span> and <span class="math display">\[
  s_g = -1 \leftrightarrow y_g = 1, \quad s_g = 1 \leftrightarrow y_g = 0.
\]</span></p>
<p>I focus on the first one as it feels more natural to me. Mathematically, we can write it as <span class="math inline">\(y_g = 0.5(s_g + 1)\)</span> or <span class="math inline">\(s_g = 1 - 2y_g\)</span>. The other relabelling can also be introduced by using an “internal” relabelling in either model (see the digression below).</p>
<div class="callout callout-style-default callout-caution callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-1-contents" aria-controls="callout-1" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Digression: internal relabelling
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-1" class="callout-1-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>Note that for the <span class="math inline">\(\{-1, 1\}\)</span> model, the mapping <span class="math inline">\(\mathbf{s}\mapsto -\mathbf{s}\)</span> for all states does not change the energy differences, once <span class="math inline">\(\mathbf{\alpha}\)</span> is changed to <span class="math inline">\(-\mathbf{\alpha}\)</span>. The interaction terms, represented by the <span class="math inline">\(\Omega\)</span> matrix, do not need to be changed.</p>
<p>In the <span class="math inline">\(\{0, 1\}\)</span> model, we can check what happens if we apply <span class="math inline">\(y_g \mapsto 1-y_g\)</span> symmetry and change <span class="math inline">\(\Theta\)</span> to <span class="math inline">\(\tilde\Theta\)</span>. We have <span class="math display">\[\begin{align*}
\tilde E(\mathbf{1} -\mathbf{y}) &amp;= \sum_{g}\sum_{g'}  \tilde \Theta_{gg'}(1-y_g)(1-y_{g'}) \\
&amp;= \sum_{g}\sum_{g'} \tilde \Theta_{gg'} y_{g}y_{g'} + \sum_{g}\sum_{g'} \tilde \Theta_{gg'} - 2\sum_{g}\left(\sum_{g'} \Theta_{g'g}  \right) y_g,
\end{align*}
\]</span></p>
<p>which looks a bit different than <span class="math display">\[
\tilde E(\mathbf{y}) = \sum_{g}\sum_{g'} \Theta_{gg'}y_gy_{g'}.
\]</span></p>
<p>However, note that:</p>
<ol type="1">
<li>The first term, i.e., interactions, does not need to be changed. We can take <span class="math inline">\(\tilde \Theta_{gg'} = \Theta_{gg'}\)</span> for all <span class="math inline">\(g\neq g'\)</span>.</li>
<li>The second term does not depend on <span class="math inline">\(\mathbf{y}\)</span>. It does not affect the energy <em>differences</em>, so that it does not change the probabilities. We can simply ignore it.</li>
<li>The last term can be incorporated in the diagonal entries of <span class="math inline">\(\Theta\)</span>, similarly as we did it with the <span class="math inline">\(\mathbf{\pi}\)</span> vector before.</li>
</ol>
<p>Hence, both models show similar symmetry. Parameterisation in terms of <span class="math inline">\(\Theta\)</span> may be more natural for biological applications, but it changes less intuitively under the relabelling symmetry (which in physics is a more natural operation than in biology).</p>
</div>
</div>
</div>
<p>The energy is given by <span class="math display">\[\begin{align*}
  \tilde E(\mathbf{y}) &amp;= \sum_{g}\sum_{g'} \Theta_{gg'} y_g y_g' \\
  &amp;= \frac{1}{4}\sum_{g}\sum_{g'} \Theta_{gg'} (1 + s_g)(1 + s_{g'}) \\
  &amp;= \mathrm{const.} + \sum_{g}\sum_{g'} \frac{\Theta_{gg'}}{4} s_g s_{g'} + \sum_{g} \left(\sum_{g'}\frac{\Theta_{gg'}}{2}\right) s_g,
\end{align*}
\]</span></p>
<p>where <span class="math inline">\(\mathrm{const.}\)</span> does not depend on <span class="math inline">\(\mathbf{y}\)</span> (or, more appropriate in this context, on <span class="math inline">\(\mathbf{s}\)</span>), so that it does not change the energy differences. and we can ignore it. From this we can read the expressions for <span class="math inline">\(\Omega\)</span> and <span class="math inline">\(\mathbf{\alpha}\)</span> vectors resulting in the same distribution. I think it is important to notice that, in many cases, <span class="math inline">\(\mathbf{\alpha}\neq 0\)</span>. In other words, to model mutational dependencies one has to use the Ising model including interactions with the external field, to introduce possible asymmetry.</p>
<p>Let’s do this exercise now in the other direction. We have <span class="math inline">\(s_g = 1 - 2y_g\)</span>, so <span class="math display">\[\begin{align*}
  E(\mathbf{s}) &amp;= \sum_g \alpha_g s_g + \sum_g \sum_{g'} \Omega_{gg'} s_g s_{g'} \\
  &amp;= \left(\sum_g \alpha_g - 2 \sum_{g} \alpha_g y_g \right) + \sum_{g}\sum_{g'} 4 \Omega_{gg'} y_g y_{g'} - \sum_g \left(\sum_{g'} 4\Omega_{gg'} \right)y_g.
\end{align*}
\]</span></p>
<p>We can happily ignore the constant term <span class="math inline">\(\sum_{g}\alpha_g\)</span> and define appropriate matrix <span class="math inline">\(\Theta\)</span> corresponding to this model. Overall, the interaction terms, i.e., <span class="math inline">\(\Omega_{gg'}\)</span> and <span class="math inline">\(\Theta_{gg'}\)</span> are equivalent up to scaling.</p>
<p>The diagonal terms seem to be more delicate, though. Let’s see what happens when <span class="math inline">\(\mathbf{\alpha} = 0\)</span>. In this case, the energy is proportional (I dropped the factor of 4) to <span class="math display">\[
  \sum_{g\neq g'} \Omega_{gg'} y_gy_g' + \sum_{g}\left( \Omega_{gg} - \sum_{g'} \Omega_{gg'} \right)y_g =
  \sum_{g\neq g'} \Omega_{gg'} y_gy_g' - \sum_{g}\left(\sum_{g'\neq g} \Omega_{gg'} \right)y_g.
\]</span></p>
<p>Note that the left hand side of the equation looks a bit wrong: it explicitly mentions the diagonal terms <span class="math inline">\(\Omega_{gg}\)</span>, even though they should not affect the energy differences! On the right hand side we see that this dependency is a mirage: the terms <span class="math inline">\(\Omega_{gg}\)</span> cancel out. In particular, this model has a strong constraint: the baseline occurence probabilities, <span class="math inline">\(\Theta_{gg}\)</span>, depend on the offdiagonal terms!</p>
<p>I don’t find this constraint very natural and I think it’s important to remember about it. I think I’d like to understand the effect of this better: in this nice paper from 2013, the analysis seems to be performed with this constraint. I wonder what happens when it is removed. In particular:</p>
<ol type="1">
<li>Are there are some interaction terms which look different?</li>
<li>Does the more flexible model fit the data better?</li>
<li>Can (penalised) pseudolikelihood still provide reliable estimates in this case? As an alternative method of training these models, one could use discrete Fisher divergence, which already <a href="../posts/discrete-intractable-likelihood.html">appeared on this blog</a>.</li>
</ol>
<p>Perhaps I should try implementing this!</p>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->



</body></html>