<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.450">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Paweł Czyż">
<meta name="dcterms.date" content="2024-04-14">
<meta name="description" content="A quick look at the differences between mixture and admixture models.">

<title>Paweł Czyż - On mixtures and admixtures</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<script src="../site_libs/quarto-html/quarto.js"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting-dark.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="dark">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>
<script async="" src="https://www.googletagmanager.com/gtag/js?id=G-PZ6W6SFE14"></script>

<script type="text/javascript">

window.dataLayer = window.dataLayer || [];
function gtag(){dataLayer.push(arguments);}
gtag('js', new Date());
 
  gtag('consent', 'default', {
    'ad_storage': 'denied',
    'analytics_storage': 'denied'
  });
gtag('config', 'G-PZ6W6SFE14', { 'anonymize_ip': true});
</script>

  <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<link rel="stylesheet" href="../styles.css">
</head>

<body class="nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg navbar-dark ">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container">
    <a class="navbar-brand" href="../index.html">
    <span class="navbar-title">Paweł Czyż</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="../index.html" rel="" target="">
 <span class="menu-text">About</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../blog.html" rel="" target="">
 <span class="menu-text">Blog</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../publications.html" rel="" target="">
 <span class="menu-text">Publications</span></a>
  </li>  
</ul>
            <div class="quarto-navbar-tools ms-auto">
</div>
          </div> <!-- /navcollapse -->
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#mixture-models" id="toc-mixture-models" class="nav-link active" data-scroll-target="#mixture-models">Mixture models</a></li>
  <li><a href="#how-expressive-are-finite-mixtures" id="toc-how-expressive-are-finite-mixtures" class="nav-link" data-scroll-target="#how-expressive-are-finite-mixtures">How expressive are finite mixtures?</a></li>
  <li><a href="#admixture-models" id="toc-admixture-models" class="nav-link" data-scroll-target="#admixture-models">Admixture models</a></li>
  <li><a href="#how-do-the-samples-look-like" id="toc-how-do-the-samples-look-like" class="nav-link" data-scroll-target="#how-do-the-samples-look-like">How do the samples look like?</a></li>
  <li><a href="#links" id="toc-links" class="nav-link" data-scroll-target="#links">Links</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">On mixtures and admixtures</h1>
</div>

<div>
  <div class="description">
    A quick look at the differences between mixture and admixture models.
  </div>
</div>


<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Paweł Czyż </p>
          </div>
  </div>
    
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">April 14, 2024</p>
    </div>
  </div>
  
    
  </div>
  

</header>

<p>Consider a binary genome vector <span class="math inline">\(Y_{n\bullet} = (Y_{n1}, \dotsc, Y_{nG})\)</span> representing at which loci a mutation has appeared. One of the simplest models is to assume that mutations appear independently, with <span class="math inline">\(\theta_g = P(Y_{ng} = 1)\)</span> representing the probability of mutation occurring. In other words, <span class="math display">\[\begin{align*}
P(Y_{n\bullet} = y_{n\bullet} \mid \theta_\bullet) &amp;= \prod_{g=1}^G P(Y_{ng}=y_{ng} \mid \theta_g ) \\
&amp;= \prod_{g=1}^G \theta_g^{Y_{ng}}(1-\theta_g)^{1-Y_{ng}}.
\end{align*}
\]</span></p>
<p>Consider the case where we observe <span class="math inline">\(N\)</span> exchangeable genomes. We will assume that they are conditionally independent given the model parameters: <span class="math display">\[
P(Y_{*\bullet} = y_{*\bullet}\mid \theta_\bullet) = \prod_{n=1}^N P(Y_{n\bullet}=y_{n\bullet}\mid \theta_\bullet).
\]</span></p>
<p>There’s an obligatory probabilistic graphical model representing our assumptions:</p>
<div class="cell" data-execution_count="1">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> daft</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> matplotlib.pyplot <span class="im">as</span> plt</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a>plt.style.use(<span class="st">"dark_background"</span>)</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="kw">class</span> MyPGM:</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a>  <span class="kw">def</span> <span class="fu">__init__</span>(<span class="va">self</span>, dpi: <span class="bu">int</span> <span class="op">=</span> <span class="dv">200</span>) <span class="op">-&gt;</span> <span class="va">None</span>:</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a>    <span class="cf">assert</span> dpi <span class="op">&gt;</span> <span class="dv">1</span></span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a>    <span class="va">self</span>.dpi <span class="op">=</span> dpi</span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a>    <span class="va">self</span>.pgm <span class="op">=</span> daft.PGM(dpi<span class="op">=</span>dpi)</span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true" tabindex="-1"></a>  <span class="kw">def</span> add_node(<span class="va">self</span>, <span class="bu">id</span>: <span class="bu">str</span>, name: <span class="bu">str</span>, x: <span class="bu">float</span>, y: <span class="bu">float</span>, observed: <span class="bu">bool</span> <span class="op">=</span> <span class="va">False</span>):</span>
<span id="cb1-12"><a href="#cb1-12" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> observed:</span>
<span id="cb1-13"><a href="#cb1-13" aria-hidden="true" tabindex="-1"></a>      params<span class="op">=</span>{<span class="st">"facecolor"</span>: <span class="st">"grey"</span>}</span>
<span id="cb1-14"><a href="#cb1-14" aria-hidden="true" tabindex="-1"></a>    <span class="cf">else</span>:</span>
<span id="cb1-15"><a href="#cb1-15" aria-hidden="true" tabindex="-1"></a>      params<span class="op">=</span>{<span class="st">"edgecolor"</span>: <span class="st">"w"</span>}</span>
<span id="cb1-16"><a href="#cb1-16" aria-hidden="true" tabindex="-1"></a>    <span class="va">self</span>.pgm.add_node(<span class="bu">id</span>, name, x, y, plot_params<span class="op">=</span>params)</span>
<span id="cb1-17"><a href="#cb1-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-18"><a href="#cb1-18" aria-hidden="true" tabindex="-1"></a>  <span class="kw">def</span> add_edge(<span class="va">self</span>, start: <span class="bu">str</span>, end: <span class="bu">str</span>):</span>
<span id="cb1-19"><a href="#cb1-19" aria-hidden="true" tabindex="-1"></a>    <span class="va">self</span>.pgm.add_edge(start, end, plot_params<span class="op">=</span>{<span class="st">"edgecolor"</span>: <span class="st">"w"</span>, <span class="st">"facecolor"</span>: <span class="st">"w"</span>})</span>
<span id="cb1-20"><a href="#cb1-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-21"><a href="#cb1-21" aria-hidden="true" tabindex="-1"></a>  <span class="kw">def</span> add_plate(</span>
<span id="cb1-22"><a href="#cb1-22" aria-hidden="true" tabindex="-1"></a>    <span class="va">self</span>,</span>
<span id="cb1-23"><a href="#cb1-23" aria-hidden="true" tabindex="-1"></a>    coords,</span>
<span id="cb1-24"><a href="#cb1-24" aria-hidden="true" tabindex="-1"></a>    label: <span class="bu">str</span>,</span>
<span id="cb1-25"><a href="#cb1-25" aria-hidden="true" tabindex="-1"></a>    shift: <span class="bu">float</span> <span class="op">=</span> <span class="fl">0.0</span>,</span>
<span id="cb1-26"><a href="#cb1-26" aria-hidden="true" tabindex="-1"></a>    label_offset: <span class="bu">tuple</span>[<span class="bu">float</span>, <span class="bu">float</span>] <span class="op">=</span> (<span class="fl">0.02</span>, <span class="fl">0.02</span>),</span>
<span id="cb1-27"><a href="#cb1-27" aria-hidden="true" tabindex="-1"></a>  ):</span>
<span id="cb1-28"><a href="#cb1-28" aria-hidden="true" tabindex="-1"></a>    <span class="co">"""</span></span>
<span id="cb1-29"><a href="#cb1-29" aria-hidden="true" tabindex="-1"></a><span class="co">    Args:</span></span>
<span id="cb1-30"><a href="#cb1-30" aria-hidden="true" tabindex="-1"></a><span class="co">      coords: [x_left, y_bottom, x_length, y_length]</span></span>
<span id="cb1-31"><a href="#cb1-31" aria-hidden="true" tabindex="-1"></a><span class="co">      label: label</span></span>
<span id="cb1-32"><a href="#cb1-32" aria-hidden="true" tabindex="-1"></a><span class="co">      shift: vertical shift</span></span>
<span id="cb1-33"><a href="#cb1-33" aria-hidden="true" tabindex="-1"></a><span class="co">    """</span></span>
<span id="cb1-34"><a href="#cb1-34" aria-hidden="true" tabindex="-1"></a>    label_offset <span class="op">=</span> (label_offset[<span class="dv">0</span>] <span class="op">*</span> <span class="va">self</span>.dpi, label_offset[<span class="dv">1</span>] <span class="op">*</span> <span class="va">self</span>.dpi)</span>
<span id="cb1-35"><a href="#cb1-35" aria-hidden="true" tabindex="-1"></a>    <span class="va">self</span>.pgm.add_plate(coords, label<span class="op">=</span>label, shift<span class="op">=</span>shift, rect_params<span class="op">=</span>{<span class="st">"edgecolor"</span>: <span class="st">"w"</span>}, label_offset<span class="op">=</span>label_offset)</span>
<span id="cb1-36"><a href="#cb1-36" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-37"><a href="#cb1-37" aria-hidden="true" tabindex="-1"></a>  <span class="kw">def</span> plot(<span class="va">self</span>):</span>
<span id="cb1-38"><a href="#cb1-38" aria-hidden="true" tabindex="-1"></a>    <span class="va">self</span>.pgm.render()</span>
<span id="cb1-39"><a href="#cb1-39" aria-hidden="true" tabindex="-1"></a>    plt.show()</span>
<span id="cb1-40"><a href="#cb1-40" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-41"><a href="#cb1-41" aria-hidden="true" tabindex="-1"></a>pgm <span class="op">=</span> MyPGM()</span>
<span id="cb1-42"><a href="#cb1-42" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-43"><a href="#cb1-43" aria-hidden="true" tabindex="-1"></a>pgm.add_node(<span class="st">"D"</span>, <span class="st">"$</span><span class="ch">\\</span><span class="st">mathcal</span><span class="sc">{D}</span><span class="st">$"</span>, <span class="dv">0</span>, <span class="dv">1</span>)</span>
<span id="cb1-44"><a href="#cb1-44" aria-hidden="true" tabindex="-1"></a>pgm.add_node(<span class="st">"theta"</span>, <span class="vs">r"$\theta_g$"</span>, <span class="dv">2</span>, <span class="dv">1</span>)</span>
<span id="cb1-45"><a href="#cb1-45" aria-hidden="true" tabindex="-1"></a>pgm.add_node(<span class="st">"Y"</span>, <span class="vs">r"$Y_</span><span class="sc">{ng}</span><span class="vs">$"</span>, <span class="dv">4</span>, <span class="dv">1</span>, observed<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb1-46"><a href="#cb1-46" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-47"><a href="#cb1-47" aria-hidden="true" tabindex="-1"></a>pgm.add_edge(<span class="st">"D"</span>, <span class="st">"theta"</span>)</span>
<span id="cb1-48"><a href="#cb1-48" aria-hidden="true" tabindex="-1"></a>pgm.add_edge(<span class="st">"theta"</span>, <span class="st">"Y"</span>)</span>
<span id="cb1-49"><a href="#cb1-49" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-50"><a href="#cb1-50" aria-hidden="true" tabindex="-1"></a>pgm.add_plate([<span class="fl">1.5</span>, <span class="fl">0.5</span>, <span class="dv">3</span>, <span class="fl">1.5</span>], label<span class="op">=</span><span class="vs">r"$g = 1, \ldots, G$"</span>, shift<span class="op">=-</span><span class="fl">0.1</span>)</span>
<span id="cb1-51"><a href="#cb1-51" aria-hidden="true" tabindex="-1"></a>pgm.add_plate([<span class="fl">2.7</span>, <span class="fl">0.55</span>, <span class="fl">1.7</span>, <span class="dv">1</span>], label<span class="op">=</span><span class="vs">r"$n=1, \ldots, N$"</span>)</span>
<span id="cb1-52"><a href="#cb1-52" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-53"><a href="#cb1-53" aria-hidden="true" tabindex="-1"></a>pgm.plot()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display">
<p><img src="mixtures-and-admixtures_files/figure-html/cell-2-output-1.png" class="img-fluid"></p>
</div>
</div>
<p>where <span class="math inline">\(\mathcal D\)</span> is the prior over the <span class="math inline">\(\theta_\bullet\)</span> vector, i.e., <span class="math inline">\(\theta_g \mid \mathcal D \sim \mathcal D\)</span> and <span class="math inline">\(\mathcal D\)</span> is supported on the interval <span class="math inline">\((0, 1)\)</span>. The simplest choice is to fix <span class="math inline">\(\mathcal D = \mathrm{Uniform}(0, 1)\)</span>, but this may be not flexible enough. Namely, different draws from the posterior on <span class="math inline">\(\theta\)</span> may look quite different than draws from <span class="math inline">\(\mathcal D\)</span>. And this discrepancy would be easy to observe once <span class="math inline">\(G\)</span> is large.</p>
<p>Hence, the simplest improvement to this model is to make <span class="math inline">\(\mathcal D\)</span> more flexible, i.e., treating it as a random probability measure with some prior on it. The simplest possibility is to use <span class="math inline">\(\mathcal D = \mathrm{Beta}(\alpha, \beta)\)</span>, where <span class="math inline">\(\alpha\)</span> and <span class="math inline">\(\beta\)</span> are given some gamma (say) priors, but more flexible models are possible (such as mixtures of beta distribution or a <a href="../posts/dirichlet-process.html">Dirichlet process</a>).</p>
<p>This model has enough flexibility to model marginal the mutation occurrence probabilities <span class="math inline">\(P(Y_{n g} = 1)\)</span>: given enough samples <span class="math inline">\(N\)</span>, the posterior on <span class="math inline">\(\theta_g\)</span> should concentrate around the average <span class="math inline">\(N^{-1}\sum_{n=1}^N Y_{ng}\)</span> (and, provided that <span class="math inline">\(\mathcal D\)</span> is flexible enough, it may concentrate near the distribution of these averages).</p>
<p>Great: this model is flexible enough to describe well the mutation probabilities. However, it’s quite rigid when it comes to mutation exclusivity and cooccurrence: <span class="math display">\[
P(Y_{n1}=1, Y_{n2} = 1) = P(Y_{n1} = 1) \cdot P(Y_{n2} = 1) = \theta_{1}\theta_{2}.
\]</span></p>
<p>We generally expect that mutations in some genes could lead to <a href="https://en.wikipedia.org/wiki/Synthetic_lethality">synthetic lethality</a>, so they would be exclusive. Also, the genes are ordered in chromosomes and <a href="https://doi.org/10.1038/s41467-020-17967-y">copy number aberrations</a> can lead to mutations being simultaneously observed in several genes (e.g., if they are in the fragment of the chromosome which is frequently lost).</p>
<p>The discrepancies between this “independent mutations” model and real data is generally easy to observe by plotting the correlation matrix between different genes. I also like looking at the the empirical distribution of the observed number of mutations in one sample, i.e., <span class="math inline">\(\sum_{g=1}^G Y_{ng}\)</span>, as the model predictis a <a href="https://en.wikipedia.org/wiki/Poisson_binomial_distribution">Poisson binomial distribution</a>, while real data may show very different behaviour.</p>
<section id="mixture-models" class="level2">
<h2 class="anchored" data-anchor-id="mixture-models">Mixture models</h2>
<p>The model above is not flexible enough to model co-occurrences and exclusivity between different mutations.</p>
<p>Imagine that we sequence <span class="math inline">\(G=3\)</span> genes which lie very close together.</p>
<p>If there is no large deletion in this chromosome, they are independently mutated with probabilities <span class="math inline">\(1\%\)</span>, <span class="math inline">\(5\%\)</span> and <span class="math inline">\(10\%\)</span>, respectively. However, if there is a large deletion, we will notice that they are all gone. Hence, it would make sense to consider two different “populations” with mutation probabilities <span class="math inline">\(\theta_{1\bullet} = (1\%, 5\%, 10\%)\)</span> and <span class="math inline">\(\theta_{2\bullet} = (100\%, 100\%, 100\%)\)</span>, where the “population” describes whether such a large deletion had place. More generally, if we consider <span class="math inline">\(K\)</span> “populations”, we can introduce sample-specific variables <span class="math inline">\(Z_n \in \{1, 2, \dotsc, K\}\)</span> and the distributions <span class="math display">\[
P(Y_{n\bullet} \mid Z_n, \{\theta_{k\bullet}\}_{k=1, \dotsc, K}) = P(Y_{n\bullet} \mid \theta_{Z_n\bullet}).
\]</span></p>
<p>We will draw this model as</p>
<div class="cell" data-execution_count="2">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>pgm <span class="op">=</span> MyPGM()</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a>pgm.add_node(<span class="st">"pi"</span>, <span class="vs">r"$\pi$"</span>, <span class="dv">0</span>, <span class="dv">0</span>)</span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a>pgm.add_node(<span class="st">"Z"</span>, <span class="st">"$Z_n$"</span>, <span class="dv">1</span>, <span class="dv">0</span>)</span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a>pgm.add_node(<span class="st">"Y"</span>, <span class="vs">r"$Y_{n\bullet}$"</span>, <span class="dv">2</span>, <span class="dv">0</span>, observed<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a>pgm.add_node(<span class="st">"D"</span>, <span class="vs">r"$\mathcal</span><span class="sc">{D}</span><span class="vs">$"</span>, <span class="dv">1</span>, <span class="dv">1</span>)</span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a>pgm.add_node(<span class="st">"theta"</span>, <span class="vs">r"$\theta_{k\bullet}$"</span>, <span class="dv">2</span>, <span class="dv">1</span>)</span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-10"><a href="#cb2-10" aria-hidden="true" tabindex="-1"></a>pgm.add_edge(<span class="st">"pi"</span>, <span class="st">"Z"</span>)</span>
<span id="cb2-11"><a href="#cb2-11" aria-hidden="true" tabindex="-1"></a>pgm.add_edge(<span class="st">"Z"</span>, <span class="st">"Y"</span>)</span>
<span id="cb2-12"><a href="#cb2-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-13"><a href="#cb2-13" aria-hidden="true" tabindex="-1"></a>pgm.add_edge(<span class="st">"D"</span>, <span class="st">"theta"</span>)</span>
<span id="cb2-14"><a href="#cb2-14" aria-hidden="true" tabindex="-1"></a>pgm.add_edge(<span class="st">"theta"</span>, <span class="st">"Y"</span>)</span>
<span id="cb2-15"><a href="#cb2-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-16"><a href="#cb2-16" aria-hidden="true" tabindex="-1"></a>pgm.add_plate([<span class="fl">0.5</span>, <span class="op">-</span><span class="fl">0.5</span>, <span class="dv">2</span>, <span class="dv">1</span>], label<span class="op">=</span><span class="vs">r"$n=1, \ldots, N$"</span>, shift<span class="op">=-</span><span class="fl">0.1</span>)</span>
<span id="cb2-17"><a href="#cb2-17" aria-hidden="true" tabindex="-1"></a>pgm.add_plate([<span class="fl">1.5</span>, <span class="fl">0.6</span>, <span class="dv">1</span>, <span class="dv">1</span>], label<span class="op">=</span><span class="vs">r"$k=1, \ldots, K$"</span>, label_offset<span class="op">=</span>(<span class="fl">0.01</span>, <span class="fl">0.2</span>))</span>
<span id="cb2-18"><a href="#cb2-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-19"><a href="#cb2-19" aria-hidden="true" tabindex="-1"></a>pgm.plot()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display">
<p><img src="mixtures-and-admixtures_files/figure-html/cell-3-output-1.png" class="img-fluid"></p>
</div>
</div>
<p>where <span class="math inline">\(\pi\in \Delta^{K-1}\)</span> is proportion vector over <span class="math inline">\(K\)</span> “populations”.</p>
<p>Due to the conditional independence structure in this model, we can also integrate out the <span class="math inline">\(Z_n\)</span> variables to get <span class="math display">\[
P(Y_n \mid \{\theta_{k\bullet}\}_{k=1, \dotsc, K}, \pi) = \sum_{k=1}^K \pi_k \, P(Y_{n\bullet} \mid \theta_{k\bullet} )
\]</span></p>
<p>which graphically corresponds to</p>
<div class="cell" data-execution_count="3">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a>pgm <span class="op">=</span> MyPGM()</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a>pgm.add_node(<span class="st">"pi"</span>, <span class="vs">r"$\pi$"</span>, <span class="dv">0</span>, <span class="dv">0</span>)</span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a>pgm.add_node(<span class="st">"Y"</span>, <span class="vs">r"$Y_{n\bullet}$"</span>, <span class="dv">2</span>, <span class="dv">0</span>, observed<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a>pgm.add_node(<span class="st">"D"</span>, <span class="vs">r"$\mathcal</span><span class="sc">{D}</span><span class="vs">$"</span>, <span class="dv">1</span>, <span class="dv">1</span>)</span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a>pgm.add_node(<span class="st">"theta"</span>, <span class="vs">r"$\theta_{k\bullet}$"</span>, <span class="dv">2</span>, <span class="dv">1</span>)</span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-9"><a href="#cb3-9" aria-hidden="true" tabindex="-1"></a>pgm.add_edge(<span class="st">"pi"</span>, <span class="st">"Y"</span>)</span>
<span id="cb3-10"><a href="#cb3-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-11"><a href="#cb3-11" aria-hidden="true" tabindex="-1"></a>pgm.add_edge(<span class="st">"D"</span>, <span class="st">"theta"</span>)</span>
<span id="cb3-12"><a href="#cb3-12" aria-hidden="true" tabindex="-1"></a>pgm.add_edge(<span class="st">"theta"</span>, <span class="st">"Y"</span>)</span>
<span id="cb3-13"><a href="#cb3-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-14"><a href="#cb3-14" aria-hidden="true" tabindex="-1"></a>pgm.add_plate([<span class="fl">0.5</span>, <span class="op">-</span><span class="fl">0.5</span>, <span class="dv">2</span>, <span class="dv">1</span>], label<span class="op">=</span><span class="vs">r"$n=1, \ldots, N$"</span>, shift<span class="op">=-</span><span class="fl">0.1</span>)</span>
<span id="cb3-15"><a href="#cb3-15" aria-hidden="true" tabindex="-1"></a>pgm.add_plate([<span class="fl">1.5</span>, <span class="fl">0.6</span>, <span class="dv">1</span>, <span class="dv">1</span>], label<span class="op">=</span><span class="vs">r"$k=1, \ldots, K$"</span>, label_offset<span class="op">=</span>(<span class="fl">0.01</span>, <span class="fl">0.2</span>))</span>
<span id="cb3-16"><a href="#cb3-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-17"><a href="#cb3-17" aria-hidden="true" tabindex="-1"></a>pgm.plot()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display">
<p><img src="mixtures-and-admixtures_files/figure-html/cell-4-output-1.png" class="img-fluid"></p>
</div>
</div>
<p>This integrated out representation is quite convenient for inference (see <a href="https://mc-stan.org/docs/stan-users-guide/finite-mixtures.html">Stan User’s guide</a>).</p>
</section>
<section id="how-expressive-are-finite-mixtures" class="level2">
<h2 class="anchored" data-anchor-id="how-expressive-are-finite-mixtures">How expressive are finite mixtures?</h2>
<p>In principle, the mixtures can be used to model an arbitrary distribution over binary vectors: consider <span class="math inline">\(K=2^G\)</span> and each <span class="math inline">\(\theta_{k\bullet}\)</span> vector to represent a different string <span class="math inline">\(0\)</span> and <span class="math inline">\(1\)</span> digits. Hence, we see that the “populations” do <em>not</em> necessarily have biological meaning. (See also Chapter 22 of <a href="http://www.stat.columbia.edu/~gelman/book/">Bayesian Data Analysis</a> or this <a href="http://bactra.org/weblog/523.html">Cosma Shalizi’s blog post</a> explaining the overinterpretation issues in factor analysis).</p>
<p>Another issue with using <span class="math inline">\(K=2^G\)</span> classes is that <span class="math inline">\(2^G\)</span> is usually <em>much</em> larger than <span class="math inline">\(N\)</span>, so that finding a suitable set of parameters <span class="math inline">\(\{\theta_{k\bullet}\}\)</span> may be tricky. We will generally prefer a smaller number of components, although using a <a href="../posts/dirichlet-process.html">Dirichlet process mixture model</a> is possible (which is quite funny, because in these models <span class="math inline">\(K=\infty\)</span> and there may be more occupied components than <span class="math inline">\(2^G\)</span>, which still result in the distribution which could be modelled with <span class="math inline">\(2^G\)</span> clusters).</p>
<p>The mixtures (with a smaller number of components than <span class="math inline">\(2^G\)</span>) are generally very popular, even if it is not always immediately clear that a model is a finite mixture: for example, <a href="https://doi.org/10.1371/journal.pcbi.1003503">this model</a> can be modelled as a mixture with <span class="math inline">\(K = G+1\)</span> components for an appropriate prior constraining the <span class="math inline">\(\theta\)</span> parameters.</p>
<p>Another example of such model are <a href="https://doi.org/10.1093/jrsssb/qkad010">Bayesian pyramids</a>: imagine that we sequence genes 1, 2, 3 on one chromosome and genes 4, 5, 6 on another chromosome, then we may consider the following idealised model employing <span class="math inline">\(K=4\)</span> “populations”:</p>
<ol type="1">
<li><span class="math inline">\(Z_n=1\)</span> means that neither chromosome is lost. Mutations in the genes arise independently.</li>
<li><span class="math inline">\(Z_n=2\)</span> means that only the first chromosome is lost, i.e., genes 1, 2, 3 are mutated. The mutations at loci 4, 5, 6 can arise independently.</li>
<li><span class="math inline">\(Z_n=3\)</span> means that the second chromosome is lost, i.e., genes 4, 5, 6 are mutated. The mutations at loci 1, 2, 3 can arise independently.</li>
<li><span class="math inline">\(Z_n=4\)</span> means that both chromosomes are lost and we observe mutations in all six genes.</li>
</ol>
<p>This model with four populations works. However, some of the parameters are tied together, as they correspond to the effects of deletion of two different chromosomes and using this structure may be important for the scalability and speed of posterior shrinkage: if we consider 10 different chromosomes, we don’t need to model <span class="math inline">\(2^10\)</span> independent clusters. More on this parameter constraints is in Section 3.1 of the original paper. (And, if you are interested in the biological applications to tumor genotypes, I’ll be showing a poster at RECOMB 2024 on this topic! You can also take a look at the <a href="https://github.com/cbg-ethz/jnotype">Jnotype</a> package).</p>
</section>
<section id="admixture-models" class="level2">
<h2 class="anchored" data-anchor-id="admixture-models">Admixture models</h2>
<p>Above we discussed that mixture models can be very expressive when they have many components (so sometimes the parameters are shared between different components).</p>
<p>Let’s think about an admixture model, which can be treated as a simplified version of the <a href="https://web.stanford.edu/group/pritchardlab/structure.html">STRUCTURE</a> model. We again assume that we have some probability vectors <span class="math inline">\(\theta_{k\bullet}\)</span> for <span class="math inline">\(k=1, \dotsc, K\)</span>. However, in this case we will call them “topics” or distinct mutational mechanisms which operate in the following way. For each gene <span class="math inline">\(Y_{ng}\)</span> there is a mechanism <span class="math inline">\(T_{ng} \in \{1, \dotsc, K\}\)</span> which is used to generate the mutation: <span class="math display">\[
  P(Y_{ng} =1 \mid T_{ng}, \{\theta_{k\bullet}\}_{k=1, \dotsc, K}) = \theta_{T_{ng}g}.
\]</span></p>
<p>The mechanisms corresponding to different genes are drawn independently within each sample, i.e., we have a sample-specific proportion vector <span class="math inline">\(\pi_n\in \Delta^{K-1}\)</span> which is used to sample the topics: <span class="math display">\[
  T_{ng} \mid \pi_n \sim \mathrm{Categorical}(\pi_n).
\]</span></p>
<p>Let’s draw the dependencies in this model:</p>
<div class="cell" data-execution_count="4">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>pgm <span class="op">=</span> MyPGM()</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a>pgm.add_node(<span class="st">"G"</span>, <span class="vs">r"$\mathcal</span><span class="sc">{G}</span><span class="vs">$"</span>, <span class="dv">0</span>, <span class="dv">1</span>)</span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a>pgm.add_node(<span class="st">"pi"</span>, <span class="vs">r"$\pi_n$"</span>, <span class="dv">0</span>, <span class="dv">0</span>)</span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a>pgm.add_node(<span class="st">"T"</span>, <span class="st">"$T_</span><span class="sc">{ng}</span><span class="st">$"</span>, <span class="dv">1</span>, <span class="dv">0</span>)</span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true" tabindex="-1"></a>pgm.add_node(<span class="st">"Y"</span>, <span class="vs">r"$Y_</span><span class="sc">{ng}</span><span class="vs">$"</span>, <span class="dv">2</span>, <span class="dv">0</span>, observed<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb4-8"><a href="#cb4-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-9"><a href="#cb4-9" aria-hidden="true" tabindex="-1"></a>pgm.add_node(<span class="st">"D"</span>, <span class="vs">r"$\mathcal</span><span class="sc">{D}</span><span class="vs">$"</span>, <span class="dv">1</span>, <span class="dv">1</span>)</span>
<span id="cb4-10"><a href="#cb4-10" aria-hidden="true" tabindex="-1"></a>pgm.add_node(<span class="st">"theta"</span>, <span class="vs">r"$\theta_{k\bullet}$"</span>, <span class="dv">2</span>, <span class="dv">1</span>)</span>
<span id="cb4-11"><a href="#cb4-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-12"><a href="#cb4-12" aria-hidden="true" tabindex="-1"></a>pgm.add_edge(<span class="st">"G"</span>, <span class="st">"pi"</span>)</span>
<span id="cb4-13"><a href="#cb4-13" aria-hidden="true" tabindex="-1"></a>pgm.add_edge(<span class="st">"pi"</span>, <span class="st">"T"</span>)</span>
<span id="cb4-14"><a href="#cb4-14" aria-hidden="true" tabindex="-1"></a>pgm.add_edge(<span class="st">"T"</span>, <span class="st">"Y"</span>)</span>
<span id="cb4-15"><a href="#cb4-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-16"><a href="#cb4-16" aria-hidden="true" tabindex="-1"></a>pgm.add_edge(<span class="st">"D"</span>, <span class="st">"theta"</span>)</span>
<span id="cb4-17"><a href="#cb4-17" aria-hidden="true" tabindex="-1"></a>pgm.add_edge(<span class="st">"theta"</span>, <span class="st">"Y"</span>)</span>
<span id="cb4-18"><a href="#cb4-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-19"><a href="#cb4-19" aria-hidden="true" tabindex="-1"></a>pgm.add_plate([<span class="op">-</span><span class="fl">0.5</span>, <span class="op">-</span><span class="fl">0.5</span>, <span class="dv">3</span>, <span class="dv">1</span>], label<span class="op">=</span><span class="vs">r"$n=1, \ldots, N$"</span>, shift<span class="op">=-</span><span class="fl">0.1</span>)</span>
<span id="cb4-20"><a href="#cb4-20" aria-hidden="true" tabindex="-1"></a>pgm.add_plate([<span class="fl">1.5</span>, <span class="fl">0.6</span>, <span class="dv">1</span>, <span class="dv">1</span>], label<span class="op">=</span><span class="vs">r"$k=1, \ldots, K$"</span>, label_offset<span class="op">=</span>(<span class="fl">0.01</span>, <span class="fl">0.2</span>))</span>
<span id="cb4-21"><a href="#cb4-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-22"><a href="#cb4-22" aria-hidden="true" tabindex="-1"></a>pgm.add_plate([<span class="fl">0.55</span>, <span class="op">-</span><span class="fl">0.45</span>, <span class="fl">1.85</span>, <span class="fl">0.8</span>], label<span class="op">=</span><span class="vs">r"$g=1, \ldots, G$"</span>, label_offset<span class="op">=</span>(<span class="fl">0.25</span>, <span class="fl">0.01</span>))</span>
<span id="cb4-23"><a href="#cb4-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-24"><a href="#cb4-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-25"><a href="#cb4-25" aria-hidden="true" tabindex="-1"></a>pgm.plot()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display">
<p><img src="mixtures-and-admixtures_files/figure-html/cell-5-output-1.png" class="img-fluid"></p>
</div>
</div>
<p>Similarly as in the mixture models, we can integrate out the latent variables: <span class="math display">\[\begin{align*}
  P(Y_{n\bullet} \mid \pi_n, \{\theta_{k\bullet}\}_{k=1, \dotsc, K}) &amp;= \sum_{T_{n\bullet}} P(Y_{n\bullet} \mid T_{n\bullet}, \{\theta_{k\bullet}\}_{k=1, \dotsc, K}) P(T_{n\bullet} \mid \pi_n ) \\
  &amp;= \sum_{T_{n\bullet}} \prod_{g=1}^G  P( Y_{ng} \mid \theta_{T_{ng}g}) P(T_{ng} \mid \pi_n) \\
  &amp;= \prod_{g=1}^G \left( \sum_{k=1}^K P(Y_{ng}\mid \theta_{kg})\, \pi_{nk} \right)
\end{align*}
\]</span></p>
<p>We see that if the proportions vector <span class="math inline">\(\pi_n\)</span> becomes <a href="https://en.wikipedia.org/wiki/One-hot">one-hot vectors</a>, then the admixture model reduces to the mixture model we have seen above. However, this model is more flexible.</p>
<p>Let’s draw the version with local variables integrated out:</p>
<div class="cell" data-execution_count="5">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>pgm <span class="op">=</span> MyPGM()</span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>pgm.add_node(<span class="st">"G"</span>, <span class="vs">r"$\mathcal</span><span class="sc">{G}</span><span class="vs">$"</span>, <span class="dv">0</span>, <span class="dv">1</span>)</span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a>pgm.add_node(<span class="st">"pi"</span>, <span class="vs">r"$\pi_n$"</span>, <span class="dv">0</span>, <span class="dv">0</span>)</span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a>pgm.add_node(<span class="st">"Y"</span>, <span class="vs">r"$Y_</span><span class="sc">{ng}</span><span class="vs">$"</span>, <span class="dv">2</span>, <span class="dv">0</span>, observed<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-8"><a href="#cb5-8" aria-hidden="true" tabindex="-1"></a>pgm.add_node(<span class="st">"D"</span>, <span class="vs">r"$\mathcal</span><span class="sc">{D}</span><span class="vs">$"</span>, <span class="dv">1</span>, <span class="dv">1</span>)</span>
<span id="cb5-9"><a href="#cb5-9" aria-hidden="true" tabindex="-1"></a>pgm.add_node(<span class="st">"theta"</span>, <span class="vs">r"$\theta_{k\bullet}$"</span>, <span class="dv">2</span>, <span class="dv">1</span>)</span>
<span id="cb5-10"><a href="#cb5-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-11"><a href="#cb5-11" aria-hidden="true" tabindex="-1"></a>pgm.add_edge(<span class="st">"G"</span>, <span class="st">"pi"</span>)</span>
<span id="cb5-12"><a href="#cb5-12" aria-hidden="true" tabindex="-1"></a>pgm.add_edge(<span class="st">"pi"</span>, <span class="st">"Y"</span>)</span>
<span id="cb5-13"><a href="#cb5-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-14"><a href="#cb5-14" aria-hidden="true" tabindex="-1"></a>pgm.add_edge(<span class="st">"D"</span>, <span class="st">"theta"</span>)</span>
<span id="cb5-15"><a href="#cb5-15" aria-hidden="true" tabindex="-1"></a>pgm.add_edge(<span class="st">"theta"</span>, <span class="st">"Y"</span>)</span>
<span id="cb5-16"><a href="#cb5-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-17"><a href="#cb5-17" aria-hidden="true" tabindex="-1"></a>pgm.add_plate([<span class="op">-</span><span class="fl">0.5</span>, <span class="op">-</span><span class="fl">0.5</span>, <span class="dv">3</span>, <span class="dv">1</span>], label<span class="op">=</span><span class="vs">r"$n=1, \ldots, N$"</span>, shift<span class="op">=-</span><span class="fl">0.1</span>)</span>
<span id="cb5-18"><a href="#cb5-18" aria-hidden="true" tabindex="-1"></a>pgm.add_plate([<span class="fl">1.5</span>, <span class="fl">0.6</span>, <span class="dv">1</span>, <span class="dv">1</span>], label<span class="op">=</span><span class="vs">r"$k=1, \ldots, K$"</span>, label_offset<span class="op">=</span>(<span class="fl">0.01</span>, <span class="fl">0.2</span>))</span>
<span id="cb5-19"><a href="#cb5-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-20"><a href="#cb5-20" aria-hidden="true" tabindex="-1"></a>pgm.add_plate([<span class="fl">0.55</span>, <span class="op">-</span><span class="fl">0.45</span>, <span class="fl">1.85</span>, <span class="fl">0.8</span>], label<span class="op">=</span><span class="vs">r"$g=1, \ldots, G$"</span>, label_offset<span class="op">=</span>(<span class="fl">0.25</span>, <span class="fl">0.01</span>))</span>
<span id="cb5-21"><a href="#cb5-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-22"><a href="#cb5-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-23"><a href="#cb5-23" aria-hidden="true" tabindex="-1"></a>pgm.plot()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display">
<p><img src="mixtures-and-admixtures_files/figure-html/cell-6-output-1.png" class="img-fluid"></p>
</div>
</div>
<p>The issue in this model is that we have <span class="math inline">\(O(KG)\)</span> parameters in the <span class="math inline">\(\theta\)</span> matrix (unless some parameter sharing is used) and additionally we have <span class="math inline">\(O(NK)\)</span> parameters of the proportion vectors. Hence, the inference in this model may be trickier to apply.</p>
</section>
<section id="how-do-the-samples-look-like" class="level2">
<h2 class="anchored" data-anchor-id="how-do-the-samples-look-like">How do the samples look like?</h2>
<p>Let’s take <span class="math inline">\(G=10\)</span> and <span class="math inline">\(K=3\)</span> and simulate some <span class="math inline">\(\theta\)</span> matrix:</p>
<div class="cell" data-execution_count="6">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> numpy <span class="im">as</span> np</span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> seaborn <span class="im">as</span> sns</span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a>G <span class="op">=</span> <span class="dv">10</span></span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a>K <span class="op">=</span> <span class="dv">3</span></span>
<span id="cb6-6"><a href="#cb6-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-7"><a href="#cb6-7" aria-hidden="true" tabindex="-1"></a>rng <span class="op">=</span> np.random.default_rng(<span class="dv">2024</span>)</span>
<span id="cb6-8"><a href="#cb6-8" aria-hidden="true" tabindex="-1"></a>theta <span class="op">=</span> np.zeros((K, G))</span>
<span id="cb6-9"><a href="#cb6-9" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> k <span class="kw">in</span> <span class="bu">range</span>(K):</span>
<span id="cb6-10"><a href="#cb6-10" aria-hidden="true" tabindex="-1"></a>  theta[k, :] <span class="op">=</span> rng.beta(<span class="dv">1</span>, <span class="dv">5</span> <span class="op">+</span> k<span class="op">**</span><span class="dv">2</span>, size<span class="op">=</span>(G,))</span>
<span id="cb6-11"><a href="#cb6-11" aria-hidden="true" tabindex="-1"></a>  theta[k, <span class="bu">min</span>(<span class="dv">3</span><span class="op">*</span>k, G):<span class="bu">min</span>(<span class="dv">3</span><span class="op">*</span>k<span class="op">+</span><span class="dv">3</span>,<span class="dv">10</span>)] <span class="op">=</span> <span class="fl">0.9</span></span>
<span id="cb6-12"><a href="#cb6-12" aria-hidden="true" tabindex="-1"></a><span class="co"># rng.uniform(size=(K, G))</span></span>
<span id="cb6-13"><a href="#cb6-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-14"><a href="#cb6-14" aria-hidden="true" tabindex="-1"></a>theta[<span class="dv">0</span>, :] <span class="op">=</span> np.sort(theta[<span class="dv">0</span>, :])</span>
<span id="cb6-15"><a href="#cb6-15" aria-hidden="true" tabindex="-1"></a>theta[<span class="op">-</span><span class="dv">1</span>, :] <span class="op">=</span> np.sort(theta[<span class="dv">1</span>, :])[::<span class="op">-</span><span class="dv">1</span>]</span>
<span id="cb6-16"><a href="#cb6-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-17"><a href="#cb6-17" aria-hidden="true" tabindex="-1"></a>fig, ax <span class="op">=</span> plt.subplots(figsize<span class="op">=</span>(<span class="dv">4</span>, <span class="dv">2</span>), dpi<span class="op">=</span><span class="dv">250</span>)</span>
<span id="cb6-18"><a href="#cb6-18" aria-hidden="true" tabindex="-1"></a>sns.heatmap(theta, vmin<span class="op">=</span><span class="dv">0</span>, vmax<span class="op">=</span><span class="dv">1</span>, cmap<span class="op">=</span><span class="st">"Greys_r"</span>, ax<span class="op">=</span>ax, xticklabels<span class="op">=</span><span class="va">False</span>, yticklabels<span class="op">=</span><span class="va">False</span>, square<span class="op">=</span><span class="va">True</span>, cbar<span class="op">=</span><span class="va">False</span>)</span>
<span id="cb6-19"><a href="#cb6-19" aria-hidden="true" tabindex="-1"></a>ax.set_xlabel(<span class="st">"Genes"</span>)</span>
<span id="cb6-20"><a href="#cb6-20" aria-hidden="true" tabindex="-1"></a>ax.set_ylabel(<span class="st">"Populations"</span>)</span>
<span id="cb6-21"><a href="#cb6-21" aria-hidden="true" tabindex="-1"></a>plt.show()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display">
<p><img src="mixtures-and-admixtures_files/figure-html/cell-7-output-1.png" class="img-fluid"></p>
</div>
</div>
<p>Let’s sample 50 vectors from each component:</p>
<div class="cell" data-execution_count="7">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>z <span class="op">=</span> np.asarray(<span class="bu">sum</span>(([k] <span class="op">*</span> <span class="dv">50</span> <span class="cf">for</span> k <span class="kw">in</span> <span class="bu">range</span>(K)), []))</span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>probs <span class="op">=</span> theta[z]</span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a>Y <span class="op">=</span> rng.binomial(<span class="dv">1</span>, probs)</span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a>fig, ax <span class="op">=</span> plt.subplots(figsize<span class="op">=</span>(<span class="dv">5</span>, <span class="dv">2</span>), dpi<span class="op">=</span><span class="dv">250</span>)</span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a>sns.heatmap(Y.T, vmin<span class="op">=</span><span class="dv">0</span>, vmax<span class="op">=</span><span class="dv">1</span>, cmap<span class="op">=</span><span class="st">"Greys_r"</span>, ax<span class="op">=</span>ax, square<span class="op">=</span><span class="va">False</span>, xticklabels<span class="op">=</span><span class="va">False</span>, yticklabels<span class="op">=</span><span class="va">False</span>, cbar<span class="op">=</span><span class="va">False</span>)</span>
<span id="cb7-7"><a href="#cb7-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-8"><a href="#cb7-8" aria-hidden="true" tabindex="-1"></a>ax.set_xlabel(<span class="st">"Samples"</span>)</span>
<span id="cb7-9"><a href="#cb7-9" aria-hidden="true" tabindex="-1"></a>ax.set_ylabel(<span class="st">"Genes"</span>)</span>
<span id="cb7-10"><a href="#cb7-10" aria-hidden="true" tabindex="-1"></a>plt.show()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display">
<p><img src="mixtures-and-admixtures_files/figure-html/cell-8-output-1.png" class="img-fluid"></p>
</div>
</div>
<p>On the other hand, if we sample from an admixture model, where <span class="math inline">\(\pi_n \sim \mathrm{Dirichlet}(\alpha, \alpha, \alpha)\)</span> for different values of <span class="math inline">\(\alpha\)</span>, we will get the following proportion vectors:</p>
<div class="cell" data-execution_count="8">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>N <span class="op">=</span> <span class="dv">150</span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a>alphas <span class="op">=</span> [<span class="fl">0.001</span>, <span class="fl">0.1</span>, <span class="fl">10.0</span>]</span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a>fig, axs <span class="op">=</span> plt.subplots(<span class="dv">3</span>, <span class="dv">1</span>, figsize<span class="op">=</span>(<span class="dv">7</span>, <span class="dv">6</span>), dpi<span class="op">=</span><span class="dv">250</span>, sharex<span class="op">=</span><span class="va">True</span>, sharey<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a>pis <span class="op">=</span> np.zeros((<span class="bu">len</span>(alphas), N, K))</span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-8"><a href="#cb8-8" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i, (alpha, ax) <span class="kw">in</span> <span class="bu">enumerate</span>(<span class="bu">zip</span>(alphas, axs)):</span>
<span id="cb8-9"><a href="#cb8-9" aria-hidden="true" tabindex="-1"></a>  pi <span class="op">=</span> rng.dirichlet(alpha <span class="op">*</span> np.ones(K), size<span class="op">=</span>N)</span>
<span id="cb8-10"><a href="#cb8-10" aria-hidden="true" tabindex="-1"></a>  index <span class="op">=</span> np.argsort( np.einsum(<span class="st">"nk,k-&gt;n"</span>, pi, np.arange(K)))</span>
<span id="cb8-11"><a href="#cb8-11" aria-hidden="true" tabindex="-1"></a>  pi <span class="op">=</span> pi[index, :]</span>
<span id="cb8-12"><a href="#cb8-12" aria-hidden="true" tabindex="-1"></a>  pis[i, ...] <span class="op">=</span> pi</span>
<span id="cb8-13"><a href="#cb8-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-14"><a href="#cb8-14" aria-hidden="true" tabindex="-1"></a>  x_axis <span class="op">=</span> np.arange(<span class="fl">0.5</span>, N <span class="op">+</span> <span class="fl">0.5</span>)</span>
<span id="cb8-15"><a href="#cb8-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-16"><a href="#cb8-16" aria-hidden="true" tabindex="-1"></a>  prev <span class="op">=</span> <span class="fl">0.0</span></span>
<span id="cb8-17"><a href="#cb8-17" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span> k <span class="kw">in</span> <span class="bu">range</span>(K):</span>
<span id="cb8-18"><a href="#cb8-18" aria-hidden="true" tabindex="-1"></a>    ax.fill_between(x_axis, prev, prev <span class="op">+</span> pi[:, k], color<span class="op">=</span><span class="ss">f"C</span><span class="sc">{</span>k<span class="op">+</span><span class="dv">4</span><span class="sc">}</span><span class="ss">"</span>)</span>
<span id="cb8-19"><a href="#cb8-19" aria-hidden="true" tabindex="-1"></a>    prev <span class="op">=</span> prev <span class="op">+</span> pi[:, k]</span>
<span id="cb8-20"><a href="#cb8-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-21"><a href="#cb8-21" aria-hidden="true" tabindex="-1"></a>  ax.spines[[<span class="st">"top"</span>, <span class="st">"right"</span>]].set_visible(<span class="va">False</span>)</span>
<span id="cb8-22"><a href="#cb8-22" aria-hidden="true" tabindex="-1"></a>  ax.set_xlabel(<span class="st">"Samples"</span>)</span>
<span id="cb8-23"><a href="#cb8-23" aria-hidden="true" tabindex="-1"></a>  ax.set_ylabel(<span class="st">"Proportions"</span>)</span>
<span id="cb8-24"><a href="#cb8-24" aria-hidden="true" tabindex="-1"></a>  ax.set_xticks([])</span>
<span id="cb8-25"><a href="#cb8-25" aria-hidden="true" tabindex="-1"></a>  ax.set_title(<span class="ss">f"$</span><span class="ch">\\</span><span class="ss">alpha$: </span><span class="sc">{</span>alpha<span class="sc">:.3f}</span><span class="ss">"</span>)</span>
<span id="cb8-26"><a href="#cb8-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-27"><a href="#cb8-27" aria-hidden="true" tabindex="-1"></a>fig.tight_layout()</span>
<span id="cb8-28"><a href="#cb8-28" aria-hidden="true" tabindex="-1"></a>plt.show()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display">
<p><img src="mixtures-and-admixtures_files/figure-html/cell-9-output-1.png" class="img-fluid"></p>
</div>
</div>
<p>Note that for <span class="math inline">\(\alpha \approx 0\)</span> the proportion vectors are close to one-hot, so that the admixture reduces to the mixture above.</p>
<p>We can generate the following samples corresponding to proportion vectors above as following:</p>
<div class="cell" data-execution_count="9">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>fig, axs <span class="op">=</span> plt.subplots(<span class="dv">3</span>, <span class="dv">1</span>, figsize<span class="op">=</span>(<span class="dv">7</span>, <span class="dv">6</span>), dpi<span class="op">=</span><span class="dv">250</span>, sharex<span class="op">=</span><span class="va">True</span>, sharey<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> alpha, pi, ax <span class="kw">in</span> <span class="bu">zip</span>(alphas, pis, axs):</span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a>  Y <span class="op">=</span> np.zeros((N, G), dtype<span class="op">=</span><span class="bu">int</span>)</span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span> n <span class="kw">in</span> <span class="bu">range</span>(N):</span>
<span id="cb9-7"><a href="#cb9-7" aria-hidden="true" tabindex="-1"></a>    T_ <span class="op">=</span> rng.multinomial(<span class="dv">1</span>, pi[n], size<span class="op">=</span>G)</span>
<span id="cb9-8"><a href="#cb9-8" aria-hidden="true" tabindex="-1"></a>    T <span class="op">=</span> T_ <span class="op">@</span> np.arange(K)</span>
<span id="cb9-9"><a href="#cb9-9" aria-hidden="true" tabindex="-1"></a>    probs <span class="op">=</span> theta[T, np.arange(G)]</span>
<span id="cb9-10"><a href="#cb9-10" aria-hidden="true" tabindex="-1"></a>    Y[n, :] <span class="op">=</span> rng.binomial(<span class="dv">1</span>, probs)</span>
<span id="cb9-11"><a href="#cb9-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-12"><a href="#cb9-12" aria-hidden="true" tabindex="-1"></a>  sns.heatmap(Y.T, vmin<span class="op">=</span><span class="dv">0</span>, vmax<span class="op">=</span><span class="dv">1</span>, cmap<span class="op">=</span><span class="st">"Greys_r"</span>, ax<span class="op">=</span>ax, square<span class="op">=</span><span class="va">False</span>, xticklabels<span class="op">=</span><span class="va">False</span>, yticklabels<span class="op">=</span><span class="va">False</span>, cbar<span class="op">=</span><span class="va">False</span>)</span>
<span id="cb9-13"><a href="#cb9-13" aria-hidden="true" tabindex="-1"></a>  ax.set_xlabel(<span class="st">"Samples"</span>)</span>
<span id="cb9-14"><a href="#cb9-14" aria-hidden="true" tabindex="-1"></a>  ax.set_ylabel(<span class="st">"Genes"</span>)</span>
<span id="cb9-15"><a href="#cb9-15" aria-hidden="true" tabindex="-1"></a>  ax.set_title(<span class="ss">f"Sparsity: </span><span class="sc">{</span>alpha<span class="sc">:.3f}</span><span class="ss">"</span>)</span>
<span id="cb9-16"><a href="#cb9-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-17"><a href="#cb9-17" aria-hidden="true" tabindex="-1"></a>fig.tight_layout()</span>
<span id="cb9-18"><a href="#cb9-18" aria-hidden="true" tabindex="-1"></a>plt.show()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display">
<p><img src="mixtures-and-admixtures_files/figure-html/cell-10-output-1.png" class="img-fluid"></p>
</div>
</div>
</section>
<section id="links" class="level2">
<h2 class="anchored" data-anchor-id="links">Links</h2>
<ul>
<li>We discussed mixture models in the <a href="../posts/dirichlet-process.html">Dirichlet process</a>.</li>
<li>There is an excellent introduction to mixture and admixture models in Jeff Miller’s <a href="http://jwmi.github.io/BMB/index.html">Bayesian methodology in biostatistics</a> course (see lectures 9–12).</li>
<li>Nicola Roberts used hierarchical Dirichlet processes (which are also an admixture model, although of a bit different kind) to study mutational patterns <a href="https://doi.org/10.17863/CAM.22674">during her PhD</a>.</li>
<li>Barbara Engelhardt and Matthew Stephens wrote a <a href="https://doi.org/10.1371/journal.pgen.1001117">nice paper</a> interpreting admixture models as matrix factorizations. Of course, mixture models (being a special case of an admixture) also fit in this framework.</li>
<li>Quantification is naturally related to admixture models. We discussed it <a href="../posts/em-gibbs-quantification.html">here</a> and <a href="../publications/bayesian-quantification.html">there</a>.</li>
</ul>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->



</body></html>